### PR TITLE
[IR] Avoid creating icmp/fcmp constant expressions

### PR DESCRIPTION
**Pull request by nikic**
Do not create icmp/fcmp constant expressions in IRBuilder etc anymore, i.e. treat them as "undesirable". This is in preparation for removing them entirely.

Part of: https://discourse.llvm.org/t/rfc-remove-most-constant-expressions/63179

### CODE REVIEW

File: llvm/test/Assembler/ConstantExprFold.ll
Comment (nikic) [inline llvm/test/Assembler/ConstantExprFold.ll:46]: There are a few cases like this where we don't canonicalize constant expressions involving icmp anymore. This is because we also refuse to create new icmp expressions even if we already had one originally.

This is fine because it only happens if there was an icmp const expr in the first place, which will be removed shortly after this lands.

File: llvm/test/Transforms/Attributor/value-simplify.ll
Comment (nikic) [inline llvm/test/Transforms/Attributor/value-simplify.ll:1420]: This change looked suspicious to me -- I was afraid that we started to simplify a null comparison through an addrspace cast, which wouldn't generally be valid. But this doesn't seem to be the case as far as ConstantFolding/InstSimplify are concerned. This is some Attributor-specific behavior and it has been flipping back and forth in this test in multiple previous changes. Maybe @jdoerfert knows what is happening here.

**Review by efriedma-quic — APPROVED** [review]
LGTM

File: clang/test/CodeGen/constantexpr-fneg.c
Comment (preames) [inline clang/test/CodeGen/constantexpr-fneg.c:12]: This looks to be a missing constant fold somewhere.  The compare should be folded to true, not a constantexpr.

Comment (preames) [reply clang/test/CodeGen/constantexpr-fneg.c:12]: Oh, ignore me.  I didn't realize this was specific testing unsimplified IR.

Comment (nikic) [reply clang/test/CodeGen/constantexpr-fneg.c:12]: I remember looking into this test a while ago (the `zext i1 true` is pretty weird, usually IRBuilder would fold that away) and the context is that one of the values here gets RAUWed and that would enable the folding. For constant expressions, replacing an operand will also re-fold the expression. For instructions, this doesn't happen. In any case, it will get folded later if optimizations are enabled.

File: clang/test/Analysis/builtin_signbit.cpp
Comment (nhaehnle) [inline clang/test/Analysis/builtin_signbit.cpp:90]: That diff is awful. We should make `update_cc_test_checks.py` use `collect_original_check_lines` :)

**Review by nhaehnle — APPROVED** [review]
LGTM

**Comment by antmox** [comment]
Hello, some bots are broken after this patch:

- https://lab.llvm.org/buildbot/#/builders/245/builds/24889
- https://lab.llvm.org/buildbot/#/builders/188/builds/46161

FAIL: Clang::weak-external.cpp

Could you look at this ?

**Comment by nikic** [comment]
@antmox Already fixed in https://github.com/llvm/llvm-project/commit/c609c04e32ef43f63a6ee54025fadf649c3247cc.

**Comment by JDevlieghere** [comment]
The LLDB bots are still failing:

https://ci.swift.org/view/all/job/llvm.org/view/LLDB/job/as-lldb-cmake/4342/
https://ci.swift.org/view/all/job/llvm.org/view/LLDB/job/lldb-cmake/2244/

**Comment by mysterymath** [comment]
This broke the Mac LLDB builders for Fuchsia as well. I'm going to revert to unblock CI.

**Comment by nikic** [comment]
Unfortunately this is a Darwin-only test, and I don't have access to any MacOS systems.

@JDevlieghere Could you please help me debug this issue? The test fails when trying to evaluate this expression: https://github.com/llvm/llvm-project/blob/58ddf3a0c6b6bbdf682ef3421d05e846a6e00527/lldb/test/API/commands/expression/weak_symbols/TestWeakSymbols.py#L37-L42

**Comment by nikic** [comment]
Okay, I managed to get access to a MacOS machine (turns out the GCC compile farm has one!) and figured out what the problem is. Basically, the presence of the constant expression meant that we previously always fell back to JIT and now we use the IR interpreter instead, which fails to handle the weak function symbol correctly. I put up a candidate fix at https://github.com/llvm/llvm-project/pull/93548.

**Comment by mikaelholmen** [comment]
Hi @nikic 

The following starts crashing with this patch:
```
opt ic2.ll -S -o /dev/null -passes=instcombine
```
It hits
```
opt: ../include/llvm/Support/Casting.h:578: decltype(auto) llvm::cast(From *) [To = llvm::TruncInst, From = llvm::Value]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
```
[ic2.ll.gz](https://github.com/user-attachments/files/15533503/ic2.ll.gz)

**Comment by nikic** [comment]
@mikaelholmen Thanks for the report, should be fixed by https://github.com/llvm/llvm-project/commit/bda8d1ad72fc72f21f6c536692594376d00db8b6.

**Comment by mikaelholmen** [comment]
> @mikaelholmen Thanks for the report, should be fixed by [bda8d1a](https://github.com/llvm/llvm-project/commit/bda8d1ad72fc72f21f6c536692594376d00db8b6).

Yep, thanks! That was fast! :)

**Pull request merged**
